<b>Привет!</b>

Меня зовут Харочкина Екатерина, и я буду проверять твой проект. Давай общаться на «ты», но если это неудобно, пожалуйста, сообщи. Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе совершенствоваться как профессионалу.

Ты подготовил замечательный проект, но есть аспекты, которые еще можно улучшить. Я буду использовать комментарии с цветовой разметкой. Пожалуйста, **не перемещай, не изменяй и не удаляй их**. Это поможет выполнить повторную проверку твоего проекта оперативнее.

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Таким комментарием выделены самые важные замечания. Без их отработки проект не будет принят.
</div>

<div class="alert alert-warning">
<b>Комментарий ревьюера №1</b> ⚠️
    
Таким комментарием выделены небольшие замечания и рекомендации, которые будет хорошо учесть.
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Таким комментарием выделены удачные решения🙂
</div>

Свои комментарии ты можешь помечать синим цветом.

<div class="alert alert-info">
<b>Комментарий студента №1</b>
    
Так ты можешь оформить свой комментарий
</div>

Давай работать над проектом в диалоге: если ты **что-то меняешь** в проекте по моим рекомендациям — **пиши об этом**. После того, как ты все доработаешь, я смогу принять твой проект!) 

# Исследование надежности заемщиков


<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Хорошая практика начинать проект с контекста, описания задачи, возможного хода исследования, а также описания имеющихся данных. Можно тут также добавить содержание с гиперссылками. Про оформление работы можно почитать в материалах курса:
    
`Полезные инструкции для учёбы - Оформление проекта - Рекомендации по выполнению проектов`
</div>

Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Разделы работы можно заканчивать промежуточными выводами
</div>

## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌ Замечание по всему разделу ответов на вопросы 
    
В каждом вопросе нужно построить общую сводную. Здесь можно воспользоваться `pivot_table()`.
    
Чтобы найти долю должников, нужно количество должников поделить на общее число заемщиков. Эту долю можно найти с помощью функции агрегации `mean`
    
Попробуй получить сводную следующего вида. В индексе столбец, который исследуем (в первом вопросе это количество детей), а столбцы: 'Всего кредитополучателей', 'Всего должников', 'Доля должников'. Учитывай, что использовать параметр columns необязательно, а в параметр aggfunc можно указывать несколько функций агрегации сразу.

Также могу посоветовать переименовать столбцы и написать функцию для построения сводных во всех вопросах.
</div>

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Найдем зависимость между количеством детей и возвратом кредита в срок. Для начала создадим новую функцию, которая создает сводные таблицы использвая метод pivot_table(). Эту функции мы в последствии будем применять и для других факторов в следующих главах исследования

In [22]:
def create_pivot(data, index_col, value_col): # объявим функцию, которая создает сводные таблицы и принимает как аргумент датафрейм (data), столбец, по которому мы группируем данные (index_col) и значения, по которым мы хотим увидеть сводную таблицу (value_col)
    pivot = data.pivot_table(index=index_col, values=value_col, aggfunc=['count','sum','mean']) # используем метод pivot_table(), куда передаем вышеизложенные аргументы, плюс, добавляем функцию aggfunc, куда передаем count для всех кредитополучателей, sum для подсчета должников и mean для расчета доли задолжности
    pivot.columns = ['Всего кредитополучателей', 'Всего должников', 'Доля должников'] # правильно назовем столбы с полученными данными 
    return pivot # выведем результат функции

In [23]:
pivot_children = create_pivot(data, 'children', 'debt') # передадим функции аргументы и расчитаем значения в сводной таблице
pivot_children

Всего кредитополучателей  Всего должников  Доля должников
children                                                           
0                            14091             1063        0.075438
1                             4808              444        0.092346
2                             2052              194        0.094542
3                              330               27        0.081818
4                               41                4        0.097561
5                                9                0        0.000000

**Вывод:** 

Рассмотрим самые большие категории - 0,1 и 2 детей. Доля задолжников между этими категориями распределилась следующим образом:

 - У заемщиков, у которых нет детей - 7.5%
 - У заемщиков с одним ребенком - 9.2%
 - У заемщиков с двумя детьми - 9.5%

Здесь прослеживается определенная закономерность, на основе которой мы можем выдвинуть следующую гипотезу - с ростом количества детей растет и задолжность по возврату кредита. На основе данной гипотезы можно утверждать, что потенциальный заемщик, у которого нет детей, охотнее возвращать кредит.

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Выявлять зависимость стоит, сравнивая самые большие, сопоставимые по размеру категории заемщиков. Какую зависимость можно выявить, сравнив категории с 0, 1 и 2 детьми?
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Хорошо
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Применим точно такой же алгоритм действий как и с колличеством детей в семье. Только в этот раз посмотрим на статус отношений клиента. Распределим клиентов по статусу отношений и в этих же категориях найдем колличество и процент клиентов, которые имели  задолженность по возврату кредитов

In [24]:
pivot_family = create_pivot(data, 'family_status', 'debt') # передадим функции аргументы и расчитаем значения в сводной таблице
pivot_family

Всего кредитополучателей  Всего должников  \
family_status                                                      
Не женат / не замужем                      2796              273   
в разводе                                  1189               84   
вдовец / вдова                              951               63   
гражданский брак                           4134              385   
женат / замужем                           12261              927   

                       Доля должников  
family_status                          
Не женат / не замужем        0.097639  
в разводе                    0.070648  
вдовец / вдова               0.066246  
гражданский брак             0.093130  
женат / замужем              0.075606

**Вывод:** 

Наблюдаемая разница в процентах совсем невелика, однако на основе полученных результатов можно выдвинуть интересную гипотезу: те клиенты, которые не состоят и не состояли в браке, выполняют свои кредитные обязательства хуже, чем те, кто состоял.

Мы можем это наблюдать в категориях "Не женат / не замужем" и "гражданский брак", где процент  задолженности составляет 9.8% и 9.3% соответственно. После них идут категории "в разводе" - 7% и "женат/замужем" - 7.6%. Меньше всего  задолженность у категории "вдовец/вдова" - 6.6%. Однако еще раз хочется отметить, что разница в процентах незначительна, поэтому, возможно, не стоит считать этот фактор (статус отношений) определяющим.

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Люди, побывавшие в официальном браке, наиболее надежны. Возможно, есть связь с возрастом. Разница может быть значима для банка
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

В третий раз применим тот же принцип и попробуем найти закономерность возврата кредита всрок с уровнем дохода клиентов банка

In [25]:
pivot_income = create_pivot(data, 'total_income_category', 'debt') # передадим функции аргументы и расчитаем значения в сводной таблице
pivot_income

Всего кредитополучателей  Всего должников  \
total_income_category                                              
A                                            25                2   
B                                          5014              354   
C                                         15921             1353   
D                                           349               21   
E                                            22                2   

                       Доля должников  
total_income_category                  
A                            0.080000  
B                            0.070602  
C                            0.084982  
D                            0.060172  
E                            0.090909

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Здесь лучше делать выводы по двум самым крупным категориям - B и С. По остальным категориям данных недостаточно для корректных выводов, статпогрешность большая
</div>

**Вывод:** 


Как и в случае с детьми, рассмотрим самые крупные представленные категории - B и C. Доля задолжников между этими категориями распределилась следующим образом:

 - У заемщиков, у которых доход варьируется от 200 тыс до 1 млн (B) - 7.1%
 - У заемщиков, у которых доход варьируется от 50 тыс до 200 тыс (С) - 8.5%

На основе полученных результатов можно выдвинуть гипотезу о том, что с ростом дохода потенциального клиента увеличивается его надежность. То есть клиенты с более высоким доходом охотнее возвращают кредиты.


<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Чем выше доход, тем ниже процент должников
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Рассмотрим фактор цели кредита.

In [26]:
pivot_purpose = create_pivot(data, 'purpose_category', 'debt') # передадим функции аргументы и расчитаем значения в сводной таблице
pivot_purpose

Всего кредитополучателей  Всего должников  \
purpose_category                                                      
операции с автомобилем                        4279              400   
операции с недвижимостью                     10751              780   
получение образования                         3988              369   
проведение свадьбы                            2313              183   

                          Доля должников  
purpose_category                          
операции с автомобилем          0.093480  
операции с недвижимостью        0.072551  
получение образования           0.092528  
проведение свадьбы              0.079118

**Вывод:** 

В очередной раз мы наблюдаем небольшую разницу в процентах между категориями. Лидерами по задолженностям являются клиенты, которые брали кредиты с целями "операции с автомобилем" и "получение образования" - 9.4% и 9.3% соответственно. Меньшей задолженностью по возврату кредита отличаются клиенты, которые берут кредиты для "операции с недвижимостью" и "проведение свадьбы" - 7.3% и 8% соответственно.

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Для округления значений можно использовать `round()`. Можно подумать над причинами такого распределения
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

Исходя из выводов, сделанных по результатам заданий выше, пропуски встречаются в столбцах days_employed и total_income. Давайте более детально рассмотрим возможные причины их возникновения. Возможные причины можно разделить на две основные группы - проблемы при сборе данных и проблемы при обработке и предоставлении данных. В первом случае, например, эти поля для заполнения данных могли быть необязательными, и клиент просто решил не предоставлять данные такого типа. Также, в некоторых случаях, сотрудники банка могли допустить ошибку при заполнении или вводе данных, что привело к пропускам в соответствующих столбцах. Эту причину также можно назвать человеческим фактором. Во втором случае, в процессе обработки данных или конвертации форматов могли возникнуть проблемы, которые привели к появлению пропусков в столбцах. Эту причину можно назвать причиной технического характера

<div class="alert alert-success">
<b>Комментарий ревьюера №1</b> ✅
    
Клиенты могли не предоставить данные. Возможно, эту информацию не требуют в некоторых случаях. Проблемы могли возникнуть на этапе загрузки и на этапе выгрузки данных. Неправильный запрос, объединение таблиц также может быть причиной.
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

В нашем случае мы использовали медианное значение для столбцов total_income и days_employed. В обоих случаях тем самым не потеряв другие важные данные, которые мы бы потеряли удалив строки с пропусками из датафрейма. Мы сохранили центральную тенденцию распределения данных по категориям типов занятости клиентов и избежали искажения, к которому мы бы пришли использовав среднее значение. Среднее значение - это среднее значение набора чисел. Медиана - это числовое значение, которое отделяет большую половину набора чисел от меньшей, то есть показывает центральное значение в выборке. Например, у нас есть данные баллов сданных студентами тестов: 10, 15, 20, 25 и 100. Среднее значение равно 34, но это недостоверное среднее значение, поскольку большинство оценок ниже 34. Для оценки успеваемости студентов может быть полезнее использовать медиану 20, поскольку она ближе к большинству оценок. Среднее значенение очень чувствительно к большим выбросам. Примером использования среднего значения может послужить измерение пульса у пацианта - в разное время пульс составлял 72, 74, 80, 76 и 78 ударов в минуту. Данные не сильно разнятся друг от друга, поэтому посчитам среднее арифметическое мы сможем точно описать среднее значения пульса у пациента - 76 ударов в минуту. 

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
    
Мы заполняли пропуски не только в total_income, но и в days_employed. Напиши, пожалуйста, в каком случае среднее и медиана различаются
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
Да, смотрим на наличие выбросов👍
</div>

### Шаг 4: общий вывод.

Целью исследования было выявление зависимостей просрочки от разных факторов/характеристик клиента – как предложенные заказчиком факторы/характеристики клиента влияют на способность потенциального заёмщика вернуть кредиту банка?

В качестве вводной информации мы получили датафрейм, который включал в себя ряд данных, описывающих клиентов. Перед тем, как перейти непосредственно к анализу данных и ответам на поставленные заказчиком вопросы, было необходимо подготовить данные к дальнейшей работе. В рамках предобработки данных были выполнены следующие шаги:

- Импортирована библиотека pandas, и данные были считаны из предложенного датафрейма.
- Изучена и проверена основная информация о датафрейме: количество строк и столбцов, типы данных и наличие пропущенных значений.
- Затем была обработка пропусков в столбцах 'total_income' и 'days_employed'.
- Для заполнения пропущенных значений в столбце 'total_income' использовался цикл for и логическая индексация. Мы заполнили пропуски медианным значением дохода для каждого типа дохода.
- Пропуски в столбце 'days_employed' также были обработаны. Для обработки пропусков в 'days_employed' мы использовали тот же метод, что и для 'total_income', перед этим убрав из выборки аномально большие значения с помощью логической индексации. Отрицательные значения были исправлены с использованием функции abs(),
 - С помощью метода .sum() мы убедились, что в датафрейме больше нет пропусков.
 - Затем для более удобной работы с числами мы изменили тип данных в столбце 'total_income' с float на int с помощью метода .astype().
 - Далее мы перешли к работе с явными дубликатами, которые обнаружили с помощью метода .duplicated().sum(), и удалили их с помощью метода .drop_duplicates().
 - Для более удобного анализа мы прибегли к категоризации данных в столбцах 'purpose' и 'total_income' с целью более удобного анализа.

В следующем этапе проекта мы перешли непосредственно к работе с данными и ответам на поставленные вопросы:

 - Есть ли зависимость между количеством детей и возвратом кредита в срок?
 - Есть ли зависимость между семейным положением и возвратом кредита в срок?
 - Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
 - Как разные цели кредита влияют на его возврат в срок?

Перед началом анализа мы создали новую функцию, которая создает сводные таблицы, используя метод pivot_table(). Проанализировав результаты, мы пришли к следующим выводам:

- С ростом количества детей увеличивается задолжность по возврату кредита. На основе данной гипотезы можно утверждать, что потенциальный заемщик, у которого нет детей, охотнее возвращать кредит.
- Клиенты, которые не состоят и не состояли в браке, выполняют свои кредитные обязательства хуже, чем те, кто состоял.
- С ростом дохода потенциального клиента увеличивается его надежность. То есть клиенты с более высоким доходом охотнее возвращают кредиты.
- Лидерами по задолженностям являются клиенты, которые брали кредиты с целями "операции с автомобилем" и "получение образования" - 9.4% и 9.3% соответственно. Меньшей задолженностью по возврату кредита отличаются клиенты, которые берут кредиты для "операции с недвижимостью" и "проведения свадьбы" - 7.3% и 8% соответственно.

Портрет надежного клиента можно описать следующим образом: клиент состоит в браке, не имеет детей, имеет доход выше 200тыс и берет кредит для операций с недвижимостью или проведения свадьбы.

Портрет ненадежного клиента можно описать следующим образом: клиент не состоит в браке, у клиента 1 и более детей, у клиента доход до 200тыс, и клиент берет займ для операций с автомобилем или получения образования.

На основе выдвинутых гипотез мы рекомендуем заказчику более тщательно собирать и обрабатывать данные клиентов и создать систему кредитного скоринга, которая бы на основе полученных данных и выдвинутых выше гипотез ранжировала клиентов по шкале надежности.

<div class="alert alert-danger">
<b>Комментарий ревьюера №1</b> ❌
 
Цель исследования состояла не в построении модели кредитного скоринга, а в выявлении зависимостей просрочки от разных факторов. Аномально большие значения 'days_employed' остались в датафрейме, их не убирали. 
    
Выводы нужно будет поправить после изменений в проекте. В первую очередь обращают внимание на выводы - это отчет перед заказчиком. Здесь нужно описать какая работа над данными была проведена и каких результатов удалось добиться. Опиши все зависимости. Рекомендованная структура выводов следующая:
    
- абзац по предобработке данных, что было сделано
    
- выявленные зависимости просрочки от разных факторов в общем виде
    
- портреты надежного и рискованного заемщика
    
- рекомендации заказчику
</div>

<div class="alert alert-success">
<b>Комментарий ревьюера №2</b> ✅
    
В качестве рекомендаций можно предложить другую категоризацию по доходам. Или исследовать другие признаки, рассчитать другие метрики (отношение дохода к сумме выплат и др.). В анкете сделать поля о доходе и стаже обязательными к заполнению, добавить автопроверку на формат вводимых пользователем данных
</div>

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
    
Ты хорошо постарался, но проект еще нуждается в доработке. Нужно построить общие сводные в каждом вопросе, проверить выводы, дополнить ответ про заполнение пропусков. Доделывай и присылай на повторную проверку.
        
Жду твоих исправлений 😊
</font></b>
</div>

<div style="border:solid purple 2px; padding: 20px"> 
<b> <font color='purple' >   
Отлично! Критических замечаний больше нет. Проект принят! 👏 Поздравляю и желаю удачи!
</font></b>
</div>